Sources/SmithyCBOR/Writer/Writer.swift

```diff
@@ -0,0 +1,224 @@
+//
+
+import AwsCommonRuntimeKit
+import Foundation
+
+@_spi(SmithyReadWrite) import protocol SmithyReadWrite.SmithyReader
+@_spi(SmithyReadWrite) import protocol SmithyReadWrite.SmithyWriter
+@_spi(Smithy) import struct Smithy.Document
+@_spi(Smithy) import protocol Smithy.SmithyDocument
+@_spi(SmithyTimestamps) import enum SmithyTimestamps.TimestampFormat
+@_spi(SmithyTimestamps) import struct SmithyTimestamps.TimestampFormatter
+
+@_spi(SmithyReadWrite)
+public final class Writer: SmithyWriter {
+    public typealias NodeInfo = String
+
+    let nodeInfo: NodeInfo
+    var cborValue: CBORType?
+    var children: [Writer] = []
+    weak var parent: Writer?
+
+    public required init(nodeInfo: NodeInfo) {
+        self.nodeInfo = nodeInfo
+    }
+
+    public required init(nodeInfo: NodeInfo, parent: Writer? = nil) {
+        self.nodeInfo = nodeInfo
+        self.parent = parent
+    }
+
+    public func data() throws -> Data {
+        let encoder = try CBOREncoder()
+        try encode(encoder: encoder)
+        let encodedBytes = encoder.getEncoded()
+        return Data(encodedBytes)
+    }
+
+    private func encode(encoder: CBOREncoder) throws {
+        if let cborValue = self.cborValue {
+            // Encode the CBOR value directly if it exists
+            encoder.encode(cborValue)
+        } else if !children.isEmpty {
+            // Encode as an indefinite-length map
+            encoder.encode(.indef_map_start)
+            for child in children {
+                guard !(child.cborValue == nil && child.children.isEmpty) else {
+                    continue
+                }
+                encoder.encode(.text(child.nodeInfo)) // Key for the child
+
+                if let cborValue = child.cborValue {
+                    encoder.encode(cborValue) // Encode the value directly
+                } else {
+                    try child.encode(encoder: encoder) // Recursively encode nested maps/arrays
+                }
+            }
+            encoder.encode(.indef_break)
+        } else {
+            // No value and no children: encode an empty map
+            encoder.encode(.indef_map_start)
+            encoder.encode(.indef_break)
+        }
+    }
+
+    public subscript(nodeInfo: NodeInfo) -> Writer {
+        if let child = children.first(where: { $0.nodeInfo == nodeInfo }) {
+            return child
+        } else {
+            let newChild = Writer(nodeInfo: nodeInfo, parent: self)
+            children.append(newChild)
+            return newChild
+        }
+    }
+
+    public func write(_ value: Bool?) throws {
+        guard let value else { return }
+        self.cborValue = .bool(value)
+    }
+
+    public func write(_ value: String?) throws {
+        guard let value else { return }
+        self.cborValue = .text(value)
+    }
+
+    public func write(_ value: Float?) throws {
+        guard let value else { return }
+        self.cborValue = .double(Double(value))
+    }
```

Comment:
Why is Float written using "bigger" type Double? Should this be changed?

Reply:
Float is a SmithyWriter type but its not a supported CBOR value type. So we convert it to bigger type Double in order to not lose precision once we're in CBOR territory